src/execution/physical_plan/plan_set_operation.cpp

```diff
@@ -106,11 +106,35 @@ unique_ptr<PhysicalOperator> PhysicalPlanGenerator::CreatePlan(LogicalSetOperati
 
 	// if the ALL specifier is not given, we have to ensure distinct results. Hence, push a GROUP BY ALL
 	if (!op.setop_all) { // no ALL, use distinct semantics
-		auto &types = result->GetTypes();
+		auto &types = result->types;
 		vector<unique_ptr<Expression>> groups, aggregates /* left empty */;
-		for (idx_t i = 0; i < types.size(); i++) {
-			groups.push_back(make_uniq<BoundReferenceExpression>(types[i], i));
+		if (op.collation_info.empty()) {
+			for (idx_t i = 0; i < types.size(); i++) {
+				groups.push_back(make_uniq<BoundReferenceExpression>(types[i], i));
+			}
+		} else { // project the collations and ordinary columns
+			vector<unique_ptr<Expression>> expressions;
+			idx_t info_idx = 0;
+			auto &info = op.collation_info;
+			for (idx_t proj_idx = 0; proj_idx < types.size(); ++proj_idx) {
+				if (info_idx < info.size() && proj_idx == info[info_idx].collation_idx) {
+					// project collation
+					auto &bound_collation_expr = info[info_idx].bound_collation_expr;
+					types[proj_idx] = bound_collation_expr->return_type;
+					expressions.push_back(std::move(bound_collation_expr));
```

Comment:
It seems that we are directly returning the collated expression. When generating a `GROUP BY` over a collated column, we should group by the collated expression and return `ANY_VALUE(col)`. This is not super obvious with `COLLATE NOCASE` because the collated expression is only lowercasing, e.g.:

```sql
select 'HELLO' collate nocase AS result union select 'HELLO';
┌─────────┐
│ result  │
│ varchar │
├─────────┤
│ hello   │
└─────────┘
```

But this becomes more obvious with ICU collations:

```sql
D SELECT 'hello' COLLATE de UNION SELECT 'HELLO';
┌────────────────────────────┐
│     'hello' COLLATE de     │
│          varchar           │
├────────────────────────────┤
│ 3832404046010901DCDCDCDCDC │
│ 383240404601090109         │
└────────────────────────────┘
```